### ISSUE TITLE
Component not re-rendering when derived atom value changes

### ISSUE THREAD

**Issue by elliotwaite** [pmndrs/jotai#1250]
I am running into a strange issue where I have a component that does not re-renderer even though the value of the derived atom that it is using has changed. I created a simplified demo that illustrates the issue:

[Demo](https://codesandbox.io/s/jotai-render-issue-e17i3j?file=/src/App.jsx) (click the "Start" button then compare what's rendered to what's being logged in the console)

Note: In this simplified demo, there is some unusual code, like using `get(derivedElapsedTimeAtom)` when I could have just used `get(elapsedTimeAtom)`, and calling `get(stateAtom)` in a place where it is not used, but in my original code, these calls made sense, and if I remove them from the demo the issue goes away.

**Comment by elliotwaite** [comment]
Here's another version of the demo with more logs that may be helpful:

[Demo (with more logs)](https://codesandbox.io/s/jotai-render-issue-with-more-logs-92p5mu?file=/src/App.jsx)

I also lowered the `TOTAL_TIME` value to 16 milliseconds so that there are fewer log messages to read through.

**Comment by dai-shi** [comment]
I think this is because you use the derived atom conditionally.
If an atom is not used _previously_, it's no longer a dependency. It's by design.

https://codesandbox.io/s/jotai-render-issue-forked-c7rb34?file=/src/App.jsx
This fixes the behavior. Although, I'm not quite sure how it's exactly causing the original behavior.

**Comment by elliotwaite** [comment]
It does seem to be related to conditionally depending on that atom, but I thought short-circuiting in derived atoms was supported. I thought that each time a derived atom was computed, its dependencies were also updated, or is that not the case?

**Comment by dai-shi** [comment]
> each time a derived atom was computed, its dependencies were also updated

that's true. so, I too am not sure the root cause. hope you can dig into it. 😄 
I wonder if it's related to strict mode, can you try production build and see if there's any difference?

**Comment by elliotwaite** [comment]
I just tested it in a production build and the issue is still occurring. I'll try to do some more digging and I'll let you know if I make any progress.

**Comment by elliotwaite** [comment]
Sorry for the delay, and had some things come up that I wanted to take care of first, but I finally found time to work on this.

I saw you opened https://github.com/pmndrs/jotai/issues/1255 which seems to be related to this issue, so perhaps you already figure out what was going on.

I was able to create a much simpler [demo](https://codesandbox.io/s/jotai-render-issue-simplified-5v6fsw?file=/src/App.jsx) (click the "Increment Count Twice" button and notice that the value of `countIsGreaterThanOne` is still `false`).

This is what seems to happen (note: I'm going to use "committed state" to mean the state of the atom in `committedAtomStateMap`):
* The committed state of `countIsGreaterThanOneAtom` is initially: `{v: false, r: 1, d: Map(1)}`
* When `countAtom` is incremented the first time:
	*  `alsoCountAtom` is invalidated.
	*  `alsoCountAtom` is added to the `pendingMap`.
	* `countIsGreaterThanOneAtom` is invalidated.
		* Its committed state becomes: `{v: false, r: 1, d: Map(1), i: 1}`
	* `countIsGreaterThanOneAtom` is added to the `pendingMap`.
		* With a `prevAtomState` value of: `{v: false, r: 1, d: Map(1)}`
	* `alsoCountAtom` is removed from the `pendingMap` and its value is recalculated.
	*  Since `alsoCountAtom` calls `get(countIsGreaterThanOneAtom)`, `countIsGreaterThanOneAtom` is also recalculated, but since its value hasn't changed (it's still `false`), it doesn't update its committed state.
	* `countIsGreaterThanOneAtom` is then removed from the `pendingMap` and its listeners (in this case, only the one `useAtomValue` reducer function) is called, however, this has no effect since its value hasn't changed.

* When `countAtom` is incremented the second time:
	*  `alsoCountAtom` is invalidated.
	*  `alsoCountAtom` is added to the `pendingMap`.
	* `countIsGreaterThanOneAtom` is invalidated.
		* However this has no effect, its committed state is still: `{v: false, r: 1, d: Map(1), i: 1}`
	* `countIsGreaterThanOneAtom` is added to the `pendingMap`.
		* This time with a `prevAtomState` value of: `{v: false, r: 1, d: Map(1), i: 1}` **(an invalidated state)**
	* `alsoCountAtom` is removed from the `pendingMap` and its value is recalculated.
	*  Since `alsoCountAtom` calls `get(countIsGreaterThanOneAtom)`, `countIsGreaterThanOneAtom` is also recalculated, and this time the value of `countIsGreaterThanOneAtom` has changed (it's now `true`) so it updates its committed state to be: `{v: true, r: 2, d: Map(1)}` **(a valid state)**
	* `countIsGreaterThanOneAtom` is then removed from the `pendingMap`, however, this time its listeners are never called because it has a committed state of `{v: true, r: 2, d: Map(1)}` and a `prevState` of `{v: false, r: 1, d: Map(1), i: 1}`, which causes the `flushPending` function to return early due to this code:
```
        if (
          prevAtomState &&
          'i' in prevAtomState &&
          atomState &&
          !('i' in atomState)
        ) {
          // We don't want to notify listeners
          // to avoid flushing a promise again (#1151)
          // and avoid extra re-renders (#1213).
          return
        }
```
Since `countIsGreaterThanOneAtom`'s listener, the `useAtomValue` reducer function, is never called, and the rendered value is never updated.

So the necessary conditions seem to be:
* The value of `countIsGreaterThanOneAtom` does not change during the first increment but does change during the second increment.
* `alsoCountAtom` is processed before `countIsGreaterThanOneAtom` in the `pendingMap`, and `alsoCountAtom` calls `countIsGreaterThanOneAtom` when being recalculated.

This leads to:
* In the first increment, the value of `countIsGreaterThanOneAtom` doesn't change, so its committed state stays invalidated, which causes the `prevAtomState` that gets added to the `pendingMap` during the second increment to be this invalidated state.
* During the second increment, the value of `countIsGreaterThanOneAtom` changes, so it updates its committed state to be a valid one, but this occurs while processing the `alsoCountAtom` value in the `pendingMap`, which occurs earlier in the `pendingMap` array. So by the time the `countIsGreaterThanOneAtom` value in the `pendingMap` is processed, it has a committed state that is valid and a `prevAtomState` that is invalidated, which matches that condition above, leading to the early return from the `flushPending` function.

I don't think I understand Jotai well enough to suggest how to resolve this issue, but I hope this information helps.

**Comment by dai-shi** [comment]
Thanks for the smaller repro! Indeed, it seems to be related with "invalidated".
I didn't know that when I opened #1255. I haven't had time for think about the solution.
We can make a test based on your repro, which ensures the future fix.

This will probably take some time as #1239, #1244, and #1255 are somewhat related.

**Comment by dai-shi** [comment]
https://codesandbox.io/s/jotai-render-issue-simplified-forked-yfuv22?file=/src/App.jsx
If I add `flushSync`, the problem is gone. So, this issue is revealed with React 18 automatic batching.

**Comment by elliotwaite** [comment]
Okay, sounds good.

I tested adding `flushSync` to the original demo, but for some reason, it didn't seem to resolve that one: [original demo with flushSync](https://codesandbox.io/s/jotai-render-issue-with-flushsync-doyk5u?file=/src/App.jsx). So either I'm not using `flushSync` correctly in that one, or there is something different between the two demos that causes `flushSync` to resolve one but not the other. Either way, I think as long as the fix resolves the smaller repro that doesn't use `flushSync`, that should also resolve the other demos.

**Comment by dai-shi** [comment]
https://codesandbox.io/s/jotai-render-issue-simplified-forked-p0lobd
Looks like v1.7.5 already fixed the issue.
We'll add a test for it #1297 and make sure #1296 won't break it.

**Comment by elliotwaite** [comment]
Nice! Sounds good.
